### ISSUE TITLE
Addition of timeUtils

### ISSUE THREAD

**Issue by rkoshak** [openhab/openhab-js#88]
This issue is for the discussion of whether or not it makes sense to add my timeUtils library to openhab-js. I can see arguments either way but I've found this one of my libraries to be so incredibly useful I think it would be of great help to all OH users.

At a high level it has the following functions:

Function | Purpose | Dependencies
-|-|-
`parseDuration` | Parses a string like "2h5s" (two hours five seconds) into a `time.Duration`. | 
`durationToDateTime` | Returns a `time.Duration` or String added to `now`. | `parseDuration`
`isISO8601` | Returns true if the passed in string matches ISO8601 format. | 
`is24Hr` | Returns true if the passed in string matches 24 hour time format (HH:MM only). |
`is12Hr` | Returns true if the passed in string matches 12 hour time format (hh:MM aa only). |
`toToday` | Takes the passed in "when", converts it to a ZonedDateTime using `toDateTime`, and moves the date to today, taking into account DST. | `toDateTime`
`toTomorrow` | Like `toToday` but changes the date to tomorrow's date. | `toDateTime`
`toYesterday` | Like `toToday` but changes the date to yesterday's date. | `toDateTime`
`betweenTimes` | Returns true if `now` or the passed in "when" is between a start or end time, handling the case where the start and end times define a period that spans midnight. | `toDateTime`, `toTomorrow`, `toYesterday`
`toDateTime` | Converts almost anything you can think of to a `time.ZonedDateTime`. See below. | `is24Hr`, `is12Hr`, `durationToDateTime`, `parseDuration`

The types handled by `toDateTime`

Type | Returns
-|-
`time.ZonedDateTime` | returns when unmodified
`java.time.ZonedDateTime` | converts to `time.ZonedDateTime` (because of parsing issue (see #54) it uses the SYSTEM timezone)
JavaScript `Date` | converts to `time.ZonedDateTime`
String | If ISO-8601 format it will eventually parse it (see #54). If 24Hr time string, creates a `time.ZonedDateTime` with that time and today's date. If 12 Hr time string, creates a `time.ZonedDateTime` with that time and today's date, if duration string adds the resultant duration to `now`
`number` or `bigint` | Adds the value as milliseconds to now.
`DateTimeType` | extracts the Java ZonedDateTime and converts it to `time.ZonedDateTime`
`PercentType` | Adds the value to `now` as seconds.
`QuantityType` | If the units  are time units, adds the time to `now`.
`DecimalType, `java.lang.Number` | Adds the value to `now` as milliseconds.

Some of the sorts of things that we can do with these.

```
if(timeUtils.betweenTimes('22:00', '09:00') // true if now is between 22:00 and 09:00, taking account that the times span midnight

if(timeUtils.betweenTimes(time.ZonedDateTime.now(), '3h', items.getItem('MyDateTime').rawstate) // true if the Item's state is between now and three hours from now

actions.ScriptExecution.createTimer(timeUtils.toDateTime('5m'), runme); // call runme in five minutes from now
```

If there is interest in my adding this, would putting it in `utils.js` be the most appropriate place or should a new namespace be created?

I don't want to go through the work or creating a PR though if we all decide this isn't appropriate for openhab-js. But if you want to look at the code you can see it at https://github.com/rkoshak/openhab-rules-tools/blob/main/timeUtils.js

**Comment by jpg0** [comment]
That all looks really useful to me. Do you think that it would be possible to attach the functions to a `ZonedDateTime` instead of leaving them as utility functions? For example, all the `toXXX` put onto `ZonedDateTime`, and the construction functions (`parseDuration`) to either add to the `ZonedDateTime` or add as `toXXX` on the source type? I ask because I personally prefer this style, and it also avoid the question of whether to create a new namespace.

**Comment by rkoshak** [comment]
I don't see why not. While the functions call each other, they don't have any sort of external dependencies (beyond needing to Java.type the openHAB State Types which I could put into the function).

Notice though there is only the one `toDateTime` function. It decided what to do based on the type of what's passed to it. 

One of the things that is powerful about that approach is the caller doesn't have to know what they have. They can just blindly call `toDateTime` and know they will get a `time.ZonedDateTime` back (or null if what was passed in can't be supported). That is probably less useful for end users but consider the other stuff in my openhab_rules_tools library. Right now anywhere I need a ZonedDateTime I can put in the one call to `toDateTime()`. If it can be converted it will work. If I separate each type into a separate function I'd have to add a dozen or so lines of code to interrogate the type to call the right conversion function. 

I wonder if `parseDuration` should be added to `time.Duration` instead of `ZonedDateTime`. That feels more consistent since it only creates a Duration anyway.

I'll see what I can do and submit it as a PR for further discussion. I usually take a hands off approach so I've never tried to attach a function to something provided by another library.  I know it's possible, just never looked into it. Now's a great time to learn me some more JavaScript. :-D

Question, would the current approach to the reference documents work or would we need to manage that some other way?

**Comment by jpg0** [comment]
> One of the things that is powerful about that approach is the caller doesn't have to know what they have. They can just blindly call toDateTime and know they will get a time.ZonedDateTime back 

Absolutely - if you add `toDateTime` to all the things that can convert, then users still don't need to know. The only difference is that they will get an exception if they try to do it on something that cannot convert (rather than a null) - which is usually better anyway.

> Question, would the current approach to the reference documents work or would we need to manage that some other way?

I don't think I can help with this - I have no idea how the docs work.

I would advise getting the opinion of @digitaldan first though - I know that he has already made changes to the js-joda code (via monkey-patching), so probably has a better idea of whether this is the right way to go or not.

**Comment by rkoshak** [comment]
Just to make sure I don't go off half cocked and that I understand:

- add a toDateTime to our item class which will cover DateTimeType, StringType, DecimalType, PercentType, and QuantityType (throwing an error in all other cases)
- monkey patch toDateTime to String to handle parseDuration, 24HrTime, 12HrTime, and in the future ISO8601 formatted strings (I'd probably add a toDateTime and a toDuration function)
- monkey patch Number to handle plus milliseconds
- monkey patch java.time.ZonedDateTime (can I do that?) to handle converting it
- monkey path Date to handle converting that

Or am I completely off base?

**Comment by jpg0** [comment]
Yes, that is what I meant. However if this is too wide a set of changes, possibly it could all live on ZonedDateTime instead, so that `ZonedDateTime.from(..)` (or similar) handles the conversions?

**Comment by rkoshak** [comment]
The hard part is already figured out. Whether I put it all in ZonedDateTime or spread it out is the same amount of work. I just want to get it right.

I'll start work on a PR and then we can go from there.

**Comment by florian-h05** [comment]
As #101 has been merged, this can be closed now??
